service/git/merge-base/merge_base.go

```diff
@@ -0,0 +1,135 @@
+package merge_base
+
+import (
+	"fmt"
+
+	"gopkg.in/src-d/go-git.v4/plumbing"
+	"gopkg.in/src-d/go-git.v4/plumbing/object"
+	"gopkg.in/src-d/go-git.v4/plumbing/storer"
+)
+
+// errIsReachable is thrown when first commit is an ancestor of the second
+var errIsReachable = fmt.Errorf("first is reachable from second")
+
+// MergeBase mimics the behavior of `git merge-base first second`, returning one
+// of the best common ancestor of the first and the second passed commits
+// The best common ancestor can not be reached from other common ancestors
+func MergeBase(
+	// REVIEWER: store param wouldn't be needed if MergeBase were part of go-git/git.Repository
+	store storer.EncodedObjectStorer,
+	first *object.Commit,
+	second *object.Commit,
+) ([]*object.Commit, error) {
+
+	secondHistory, err := ancestorsIndex(first, second)
+	if err == errIsReachable {
+		return []*object.Commit{first}, nil
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	var res []*object.Commit
+	inSecondHistory := isInIndexCommitFilter(secondHistory)
+	// REVIEWER: store argument wouldn't be needed if this were part of go-git/plumbing/object package
+	resIter := NewFilterCommitIter(store, first, &inSecondHistory, &inSecondHistory)
+	err = resIter.ForEach(func(commit *object.Commit) error {
+		res = append(res, commit)
+		return nil
+	})
+
+	return independents(res, 0)
+}
+
+// ancestorsIndex returns a map with the ancestors of the first commit if the
+// second one is not one of them. It returns errIsReachable if the second one is
+// ancestor, or another error if the history is not transversable
+func ancestorsIndex(first, second *object.Commit) (map[plumbing.Hash]bool, error) {
+	if first.Hash.String() == second.Hash.String() {
+		return nil, errIsReachable
+	}
+
+	secondHistory := map[plumbing.Hash]bool{}
+	secondIter := object.NewCommitIterBSF(second, nil, nil)
+	err := secondIter.ForEach(func(commit *object.Commit) error {
+		if commit.Hash == first.Hash {
+			return errIsReachable
+		}
+
+		secondHistory[commit.Hash] = true
+		return nil
+	})
+
+	if err == errIsReachable {
+		return nil, errIsReachable
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return secondHistory, nil
+}
+
+// independents returns a subset of the passed commits, that are not reachable
+// from any other. Since A can be not reachable from B, but B can be reached
+// from A, it must be checked both directions, traversing the history of all
+// passed commits, and removing those that are reachable from any traversed history.
+// Every time a history is traversed, and the ancestor found have been removed
+// from the subset of independents commits, the function is called recursively
+// with the new subset.
+func independents(commits []*object.Commit, start int) ([]*object.Commit, error) {
+	if len(commits) == 1 {
+		return commits, nil
+	}
+
+	res := commits
+	for i := start; i < len(commits); i++ {
+		from := commits[i]
+		fromHistoryIter := object.NewCommitIterBSF(from, nil, nil)
```

Comment:
I don't dully understand the algorithm here so this is only a comment for future improvements. It seems to me that this is walking the commit tree almost fully several times. This may be ok for small repositories but it may have a penalty for bigger ones like kubernetes. Testing it with a PR in kubernetes my be be interesting.

Reply:
Thanks @jfontan for your review!!!
Let's see if :point_down: this helps you to understand the algorithm.

Yes, the history can potentially be walked as many times as candidates provided for `Independents`.
But every time a candidate is found as an ancestor of any other, it is removed from the candidates list, so it won't be considered anymore.
The order in which you request the candidates, affects how many times the history is traversed.
In the worst scenario, the history will be traversed as many times as orphan branches are included in the candidates to be analyzed.

given this history:
```
          o----o----B~2---B^---B
         /         /
--A~4---o---A~2---o---A

---o---C^---C
```

then:
- **(A)** `Independents(B, B^)`; `1step`; walks the history only once, because starting by `B`, it reaches in one step `B^`; once `B^` is reached, it is removed from the candidates list, and since there only one candidate left (`B`), it is returned with no more steps. 
- **(B)**`Independents(B, B^, B~2)`; `2steps`; walks only once (as in the previous example), but it requires two steps to reach `B^` and then `B~2`
- **(C)**`Independents(B, A~4)`; `8steps`; walks the history only once (as in the previous example), but it requires many steps to reach `A~4`
- **(D)**`Independents(B^, B)`; `1full + 1step`; walks the history twice, because first iteration walking from `B^` does not reach `B` after a full walk from `B^`; then in the second iteration walking from `B` it is the same case as **(A)**, so it is returned `A` with only one extra step
- **(E)** `Independents(B, A)`; `2full`; walks the history twice; because during first walk from `B` it won't be reached `A`, and during the second walk from `A`, it won't be reached `B`
- **(F)** `Independents(B, A, B^)`; `2full`; walks the history twice; first time, it will be found `B^` as an ancestor, so it will be removed from candidates list; then it will be the same case than **E**
- **(G)**  `Independents(B, B^, B~2, A, A~2, A~4, C, C^)`; `3full`; because during the first history walk from `B` it will be removed `B^`, `B~2`, `A~2`, `A~4` from the candidates list; then the second iteration will consider only `B`, `A`, `C` and `C^`, then, walking from `A`, it won't find any ancestor in the candidates list; the third iteration, starting from `C` will find (and remove) `C^` from the list, so the result will be only `A`, `B` and `C`.

possible optimizations:
- sort the candidates by `Committer.When desc` before start processing them, so most examples of **(D)** (`Independents(B^, B)`) will be converted into **(A)** cases (`Independents(B,B^)`); it won't work when `Committer.When` does not reflect the topological order (in such cases, it will cause the problem tried to be solved, but I think it would be a "strange" case)
- store all traversed commits and exclude them in the next history walk. It will make that **(E)** `Independents(B, A)` will require only `1full+2steps` instead of `2full` because the first iteration (from `B`) will be used to shortcut the second one (from `A`) because since `A~2`, `A` shares `B` history. 
- minor fix: replace `s/commits/res` from [`line 92`](https://github.com/src-d/lookout/pull/530/files#diff-e94421aaafbf6f91a66bfeb277201f81R92) to exclude already excluded ancestors.

and thanks to your question, I spotted a bug:
- replace `start` from [`line 110`](https://github.com/src-d/lookout/blob/d39d6b49f15954136e6b4e20133aa23094a33745/service/git/merge-base/merge_base.go#L110) with `find(from, res) + 1)` (pseudocode), because otherwise the iteration starts from the same place as many times as an ancestor is found from the `start` position, what is clearly too much, and does not reflect the desired (and described) behavior.